shared/chat/conversation/list-area/normal/index.desktop.js

```diff
@@ -3,94 +3,82 @@
 import * as Virtualized from 'react-virtualized'
 import * as React from 'react'
 import Message from '../../messages'
+import SpecialTopMessage from '../../messages/special-top-message'
+import SpecialBottomMessage from '../../messages/special-bottom-message'
 import {ErrorBoundary} from '../../../../common-adapters'
 import clipboard from '../../../../desktop/clipboard'
 import debounce from 'lodash/debounce'
 import {globalColors, globalStyles} from '../../../../styles'
 
 import type {Props} from '.'
 
-type State = {
-  isLockedToBottom: boolean,
-  listRerender: number,
-}
-
 const lockedToBottomSlop = 20
 
-class Thread extends React.Component<Props, State> {
+class Thread extends React.Component<Props> {
   _cellCache = new Virtualized.CellMeasurerCache({
     fixedWidth: true,
-    keyMapper: (rowIndex: number) => this.props.messageOrdinals.get(rowIndex),
+    keyMapper: (index: number) => {
+      const itemCountIncludingSpecial = this._getItemCount()
+      if (index === itemCountIncludingSpecial - 1) {
+        return 'specialBottom'
+      } else if (index === 0) {
+        return 'specialTop'
+      } else {
+        const ordinalIndex = index - 1
+        return this.props.messageOrdinals.get(ordinalIndex)
+      }
+    },
   })
 
   _list: any
-  _keepIdxVisible: number = -1
-  _lastRowIdx: number = -1
-  // ScrolltoRow sometimes triggers load more. This'll all go away when we ditch react-virtualized. for now this fixes
-  // if you edit your last message that'l cause a load more sometimes due to scrolltop being 0 incorrectly
-  _ignoreScrollUpTill: number = 0
-
-  state = {
-    isLockedToBottom: true,
-    listRerender: 0,
-    selectedMessageKey: null,
-  }
-
-  componentDidUpdate(prevProps: Props, prevState: State) {
-    // Force a rerender if we passed a row to scroll to. If it's kept around the virutal list gets confused so we only want it to render once basically
-    if (this._keepIdxVisible !== -1) {
-      this.setState(prevState => ({listRerender: prevState.listRerender + 1})) // eslint-disable-line react/no-did-update-set-state
-      this._keepIdxVisible = -1
-    }
-    this._lastRowIdx = -1 // always reset this to be safe
+  // If we should stick to the bottom
+  _isLockedToBottom: boolean = true
```

Comment:
i moved this from a state to a var. it only matters if we actually have different data (which causes a render anyways) and if we don't do this as you lock/unlock it would rerender (for no real reason)